### PR TITLE
Adds a way to automatically delete inactive ships

### PR DESCRIPTION
**Pull request by SuperNovaa41**
Adds a system to check upon the status of the crew upon each member's death.

Closes https://github.com/Jackriip/Voidcrew-LRP/pull/129

If the crew is all dead (or all SSD, with a 10 minute buffer time), the ship will either be deleted (if its flying through space), or turned into a ruin upon the planet its on (or the space its docked in space).
This will:
    - Open up more ship slots
    - Add mothballed ships onto planets that can be explored and harvested of its loot
    - STOP THE ENDLESS HOARD OF PILLS FUCKING OVER EVERYONE ELSE
Inactive ships on planets will be mothballed after 10 minutes, flying ships will be fully deleted after 2 minutes

Just to clarify this DOESNT delete an entirely SSD ship, a crewmate has to die for this check to occur

This deletes the cryopod upon mothballing. I was considering messing with the engines too.. but i'm not quite sure yet. Let me know!

:cl:
Add: Inactive ships will now be automatically deleted, or turned into ruins, to prevent clogged up ship slots and pills ruling the server.
/:cl:

### CODE REVIEW

**Comment by theta006** [comment]
if you kill a crew and loot the ship what happens to the people on the ship when it despawns

**Comment by SuperNovaa41** [comment]
> if you kill a crew and loot the ship what happens to the people on the ship when it despawns

i gotta test that, also i should add a timer because it really shouldnt delete instantly.
also, if its docked on a planet im going to turn it into a ruin

**Comment by theta006** [comment]
also how would you keep hijacked ships (incase they fucked your ship)

**Comment by SuperNovaa41** [comment]
> also how would you keep hijacked ships (incase they fucked your ship)

Fairly certain hijacking another ship is against the rules so... not too worried about something like that

**Comment by theta006** [comment]
> > also how would you keep hijacked ships (incase they fucked your ship)
> 
> Fairly certain hijacking another ship is against the rules so... not too worried about something like that

what do you do if another crew fucked over your ship and you killed them all

File: code/modules/shuttle/shuttle.dm [hunk 1]
```diff
@@ -453,6 +453,36 @@
...
+
+	var/area/underlying_area = GLOB.areas_by_type[underlying_area_type]
+	if(!underlying_area)
+		underlying_area = new underlying_area_type(null)
+
+	for(var/i in 1 to old_turfs.len)
+		var/turf/oldT = old_turfs[i]
+		if(!oldT || !istype(oldT.loc, area_type))
+			continue
+		var/obj/machinery/cryopod/pod = locate() in oldT.contents
```
Comment (SuperNovaa41) [inline code/modules/shuttle/shuttle.dm:480]: This is probably poopy and should be some sort of callback, but im not quite sure where i would put it, if anyone has any ideas lmk

**Comment by githubuser4141** [comment]
Could the timer on flying or space-docked ships be upped to 30 minutes or more, cause they aren't really compromising gameplay save for map clutter or lag, maybe. Please don't have flying ships deleted after 2 minutes, that's FAR too aggressive when they aren't causing issues for anyone, and it'll really screw over people with worse connections who disconnect frequently (sometimes you need to re-download resources which takes 5-20 minutes.)


Would it be practical to make cryopods/cryo consoles actual computers and machinery so they could be printed and attached to a ship's helm PC via multitool?

**Comment by SuperNovaa41** [comment]
> Could the timer on flying or space-docked ships be upped to 30 minutes or more, cause they aren't really compromising gameplay save for map clutter or lag, maybe. Please don't have flying ships deleted after 2 minutes, that's FAR too aggressive when they aren't causing issues for anyone, and it'll really screw over people with worse connections who disconnect frequently (sometimes you need to re-download resources which takes 5-20 minutes.)

I don't think you quite know how the system works here, the ship is only deleted, or even CONSIDERED to be deleted upon a crew member's death. for small crews (and single such as pills) this most likely won't even be an issue. However, if it DOES prove to be an issue, I'm fine with upping that number. 

> compromising gameplay save for map clutter or lag

for this specifically, its not map clutter or lag we're worried about, it's the fact that people rush pills and then die and then buy more pills, etc. There are only so many ship slots, and they tend to all be filled up by abandoned pills.

> Would it be practical to make cryopods/cryo consoles actual computers and machinery so they could be printed and attached to a ship's helm PC via multitool?

they actually already ARE actual machines and computers, they just uhh aren't printable or deconstructable AFAIK

**Comment by githubuser4141** [comment]
> > Could the timer on flying or space-docked ships be upped to 30 minutes or more, cause they aren't really compromising gameplay save for map clutter or lag, maybe. Please don't have flying ships deleted after 2 minutes, that's FAR too aggressive when they aren't causing issues for anyone, and it'll really screw over people with worse connections who disconnect frequently (sometimes you need to re-download resources which takes 5-20 minutes.)
>
> I don't think you quite know how the system works here, the ship is only deleted, or even CONSIDERED to be deleted upon a crew member's death. for small crews (and single such as pills) this most likely won't even be an issue. However, if it DOES prove to be an issue, I'm fine with upping that number.

That's a little better, but god damn 2 minutes is still too short. Respawn timers are 2 and a half minutes, so that makes it impossible to respawn on your ship if it's solo-operated which is a real shame if you die to wildlife or lag (like: walking out the hangar from lag). I've bought ships before and ran them solo (nobody else joined) for the entire round, imagine if you died to lag and 2 1/2 hours of progress went out the window because the respawn timer is longer then the delete timer.

**Comment by SuperNovaa41** [comment]
oh fuck i completely forgot respawn timers were that long. Yeah ill up it to 10 minutes

File: code/modules/overmap/ships/simulated_ship_data.dm [hunk 2]
```diff
@@ -15,11 +22,74 @@
...
+		return // ssd guy came back
+
+	switch (state)
+		if (OVERMAP_SHIP_FLYING)
+			addtimer(CALLBACK(src, .proc/destroy_ship), SHIP_DELETE)
+		if (OVERMAP_SHIP_UNDOCKING)
+			addtimer(CALLBACK(src, .proc/destroy_ship), SHIP_DELETE)
+		if (OVERMAP_SHIP_ACTING)
+			// delete it because this is somewhat ambiguous (but they are technically flying here)
+			destroy_ship()
```
Comment (SuperNovaa41) [inline code/modules/overmap/ships/simulated_ship_data.dm]: This should be a timer

**Review by SuperNovaa41 — COMMENTED** [review]
note to self: add admin messages for this + ONE MORE CHECK MUST BE MADE BEFORE FINALIZING THE ACTION

**Pull request merged**